src/Disks/ObjectStorages/MetadataStorageFromPlainRewritableObjectStorage.cpp

```diff
@@ -109,15 +105,20 @@ void MetadataStorageFromPlainRewritableObjectStorage::load(bool is_initial_load)
             LOG_WARNING(log, "Legacy layout is likely used for disk '{}'", object_storage->getCommonKeyPrefix());
         }
 
-        if (!has_metadata)
+        if (!has_data && !has_metadata)
         {
             LOG_DEBUG(log, "Loaded metadata (empty)");
+            fs_tree->apply(std::move(remote_layout));
             return;
         }
     }
 
     try
     {
+        /// Root folder is a special case. Files are stored as /__root/{file-name}.
+        for (auto iterator = object_storage->iterate(std::filesystem::path(object_storage->getCommonKeyPrefix()) / ROOT_FOLDER_TOKEN, 0); iterator->isValid(); iterator->next())
+            remote_layout[""].file_names.insert(fs::path(iterator->current()->getPath()).filename());
```

Comment:
To confirm, do I understand correctly that the change does not require the presence of the `/__root/` folder in the object storage?
It looks that `remote_layout[""] = DirectoryRemoteInfo{ROOT_FOLDER_TOKEN, "fake_etag", 0, {}};` above ensures backward compatibility, that the data on the existing disks will be loaded into the new directory tree structure.

Reply:
It makes sense to add a compatibility integration test to have confidence in layout changes across different CH versions – not as a part of this PR. I can do it next week if nothing urgent comes up.